From 6a6d27bb9afc4fbc0da7d600d2ffd25b51c738a1 Mon Sep 17 00:00:00 2001
From: PyScotch <pyscotch@noreply.invalid>
Date: Tue, 13 Oct 2026 13:40:21 +0000
Subject: [PATCH 02/11] Bugfix: flag the arrays of 3D grid graphs as freeable
 only once the graph is built [report R. Perry]

`dgraphBuildGrid3D()` set the `DGRAPHFREETABS` flag before calling
`dgraphBuild2()`. When the build failed after the arrays had been
attached to the graph, e.g. on an allocation failure in the vertex
label translation of `dgraphBuild3()`, the routine freed them and
`dgraphExit()` freed them again. Moreover, the flags were assigned,
not or'ed, which discarded the flags already set in the structure.

Set the flags after a successful build, as `dgraphBuildHcub()` does.

Refs: DGR-18
---
 src/libscotch/dgraph_build_grid3d.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/libscotch/dgraph_build_grid3d.c b/src/libscotch/dgraph_build_grid3d.c
index 730e359..aa2a53e 100644
--- a/src/libscotch/dgraph_build_grid3d.c
+++ b/src/libscotch/dgraph_build_grid3d.c
@@ -45,7 +45,7 @@
 /**                # Version 5.1  : from : 05 jun 2010     **/
 /**                                 to   : 06 jun 2010     **/
 /**                # Version 7.0  : from : 17 jan 2023     **/
-/**                                 to   : 28 jul 2024     **/
+/**                                 to   : 13 oct 2026     **/
 /**                                                        **/
 /************************************************************/
 
@@ -473,8 +473,6 @@ const Gnum                  flagval)              /* Grid type        */
   }
   vertloctax[vertlocnum] = edgelocnum;            /* Mark end of local vertex array */
 
-  grafptr->flagval = (DGRAPHFREETABS | DGRAPHVERTGROUP | DGRAPHEDGEGROUP); /* All arrays will be freed on exit */
-
   if (dgraphBuild2 (grafptr, baseval,             /* Build distributed graph */
                     vertlocnbr, vertlocnbr, vertloctax, vertloctax + 1, veloloctax, velolocsum, NULL, vlblloctax,
                     edgelocnum - baseval, edgelocsiz, datadat.edgeloctax, NULL, datadat.edloloctax, degrglbmax) != 0) {
@@ -483,5 +481,7 @@ const Gnum                  flagval)              /* Grid type        */
     return  (1);
   }
 
+  grafptr->flagval |= DGRAPHFREETABS | DGRAPHVERTGROUP | DGRAPHEDGEGROUP; /* Arrays created by the routine itself */
+
   return (0);
 }
-- 
2.47.3

